### PR TITLE
Add tags option during enroll/install

### PR DESCRIPTION
**Pull request by narph**
## What is the problem this PR solves?

Issue https://github.com/elastic/elastic-agent/issues/149

Add ability to set tags during installation and enrolment agent
so that different identical deployments of service can be differentiated in the dataset

The specific use case is described in: https://discuss.elastic.co/t/fleet-managed-elastic-agent-environment-deployment-name-custom-field/282565/6

```
sudo ./elastic-agent install -f --url=https://<feet manager>:443 --enrollment-token=<token> --tag=staging01 linux
or in case, the agent is already installed via package manager

elastic-agent enroll -f --url=https://<feet manager>:443 --enrollment-token=<token> --tag=staging01 linux
```

Related https://github.com/elastic/fleet-server/pull/1350

Desired behavior:

```
 .\elastic-agent.exe enroll  --url=https://192.0.2.1:8220 --fleet-server-es=http://localhost:9200  --fleet-server-service-token=AAEAAWVsYXN0aWMvZmxlZXQtc2VydmVyL3Rva2VuLTE2NTA5NTg3Mjc0OTU6TmVHbEx4WWtUaVNETHNmNmNPV09lZw  --fleet-server-policy=fleet-server-policy --insecure --tag this-is-fleet,windows
```

```
{
        "_index": ".fleet-agents-7",
        "_id": "d1c66b32-72ed-4ab3-a36c-9660a5cb7125",
        "_score": null,
        "_source": {
          "access_api_key_id": "TW7SZIAB2-L_WJKcJz9m",
          "action_seq_no": [
            -1
          ],
          "active": true,
          "agent": {
            "id": "d1c66b32-72ed-4ab3-a36c-9660a5cb7125",
            "version": "8.3.0"
          },
          "enrolled_at": "2022-04-26T07:42:15Z",
          "local_metadata": {
            "elastic": {
              "agent": {
                "build.original": "8.3.0-SNAPSHOT (build: 2312423496161dd6f05148d2df0941d2c74570c8 at 2022-04-25 13:23:01 +0000 UTC)",
                "id": "d1c66b32-72ed-4ab3-a36c-9660a5cb7125",
                "log_level": "info",
                "snapshot": true,
                "upgradeable": true,
                "version": "8.3.0"
              }
            },
            "host": {
              "architecture": "x86_64",
              "hostname": "DESKTOP-K76UDQL",
              "id": "bc395ab9-1a1a-4312-ad0b-a05cceaf78f1",
              "ip": [
                "fe80::1d58:b17:4161:9414/64",
                "169.254.148.20/16",
                "2a02:a446:f25b:1:558d:72b2:74b6:7c67/64",
                "fe80::558d:72b2:74b6:7c67/64",
                "192.168.2.26/24",
                "fe80::d135:57b1:9b7f:e10a/64",
                "169.254.225.10/16",
                "fe80::6901:1ee0:60dd:c986/64",
                "169.254.201.134/16",
                "fe80::f09a:4481:bc48:e3e3/64",
                "192.168.21.110/24",
                "fe80::fc98:6ed8:6d25:3d98/64",
                "169.254.61.152/16",
                "::1/128",
                "127.0.0.1/8",
                "fe80::b54c:5acc:d25c:75e9/64",
                "172.23.192.1/20",
                "fe80::3186:c6c:cd37:92b8/64",
                "172.18.224.1/20"
              ],
              "mac": [
                "00:ff:6d:ef:82:d6",
                "cc:48:3a:b9:87:c0",
                "9e:b6:d0:3e:a8:c4",
                "ae:b6:d0:3e:a8:c4",
                "9c:b6:d0:3e:a8:c4",
                "28:c2:1f:52:ff:bf",
                "00:15:5d:a0:47:13",
                "00:15:5d:ff:9c:07"
              ],
              "name": "DESKTOP-K76UDQL"
            },
            "os": {
              "family": "windows",
              "full": "Windows 10 Pro(10.0)",
              "kernel": "10.0.22000.613 (WinBuild.160101.0800)",
              "name": "Windows 10 Pro",
              "platform": "windows",
              "version": "10.0"
            }
          },
          "policy_id": "fleet-server-policy",
          "tags": [
            "this-is-fleet",
            "windows"
          ],
          "type": "PERMANENT"
        },
        "sort": [
          1650958935000
        ]
      }
```

```
./elastic-agent.exe container -h

....
The following environment variables are provided as a convenience to prevent a large number of environment variable to
be used when the same credentials will be used across all the possible actions above.

  ELASTICSEARCH_HOST - elasticsearch host [http://elasticsearch:9200]
  ELASTICSEARCH_USERNAME - elasticsearch username [elastic]
  ELASTICSEARCH_PASSWORD - elasticsearch password [changeme]
  ELASTICSEARCH_CA - path to certificate authority to use with communicate with elasticsearch
  KIBANA_HOST - kibana host [http://kibana:5601]
  KIBANA_FLEET_USERNAME - kibana username to enable Fleet [$ELASTICSEARCH_USERNAME]
  KIBANA_FLEET_PASSWORD - kibana password to enable Fleet [$ELASTICSEARCH_PASSWORD]
  KIBANA_CA - path to certificate authority to use with communicate with Kibana [$ELASTICSEARCH_CA]
  ELASTIC_AGENT_TAGS - user provided tags for the agent
....
```

### CODE REVIEW

**Comment by blakerouse** [comment]
Does this need to also have ENV variables for the container?

**Comment by ph** [comment]
@blakerouse Yes, I think we need it for container

File: internal/pkg/agent/cmd/enroll_cmd.go [hunk 1]
```diff
@@ -498,6 +499,7 @@ func (c *enrollCmd) enroll(ctx context.Context, persistentConfig map[string]inte
 		Metadata: fleetapi.Metadata{
 			Local:        metadata,
 			UserProvided: c.options.UserProvidedMetadata,
+			Tags:         c.options.Tags,
 		},
```
Comment (ph) [inline internal/pkg/agent/cmd/enroll_cmd.go:503]: We should probably run all the tags through `strings.TrimSpace` and validate we aren't sending any empty string?

File: internal/pkg/agent/cmd/container.go [hunk 2]
```diff
@@ -126,14 +126,15 @@ be used when the same credentials will be used across all the possible actions a
   KIBANA_FLEET_USERNAME - kibana username to enable Fleet [$ELASTICSEARCH_USERNAME]
   KIBANA_FLEET_PASSWORD - kibana password to enable Fleet [$ELASTICSEARCH_PASSWORD]
   KIBANA_CA - path to certificate authority to use with communicate with Kibana [$ELASTICSEARCH_CA]
+  ELASTIC_AGENT_TAGS - user provided tags for the agent
```
Comment (ph) [inline internal/pkg/agent/cmd/container.go:129]: Can you add the expected format, is this a comma-separated list?

**Review by ph — CHANGES_REQUESTED** [review]
@narph is there any unit test that could be written here? Do we have the required mechanic in place to write them?

File: internal/pkg/agent/cmd/container.go (hunk 2 above)
Comment (blakerouse) [reply internal/pkg/agent/cmd/container.go:129]: Should it be prefixed with `ELASTIC_AGENT`? Are the tags useful in standalone mode? If not I think that `FLEET_TAGS` would be better. As its the tags that will be show for this Elastic Agent in Fleet.

Comment (ph) [reply internal/pkg/agent/cmd/container.go:129]: We are starting to have so many tags with a different prefixes, @nimarezainia I would like your opinion on this.

Comment (narph) [reply internal/pkg/agent/cmd/container.go:129]: I thought tags apply for enrolled agents only (can I get a confirmation here @nimarezainia?)
I chose the `ELASTIC_AGENT_ ` naming as the tag applies to the specific agent enrolled and not to the Fleet Server, agent is the subject receiving the tags, that said, I am open to any naming.

**Comment by narph** [comment]
> @narph is there any unit test that could be written here? Do we have the required mechanic in place to write them?

I've created a small unittest, we should update the e2e tests to validate the full flow.

File: internal/pkg/agent/cmd/container.go (hunk 2 above)
Comment (nimarezainia) [reply internal/pkg/agent/cmd/container.go]: these tags only apply to enrolled agents (managed agents) "ELASTIC_AGENT_TAGS" sounds right to me per @narph explanation.

**Comment by nimarezainia** [comment]
@narph thank you for taking care of this issue.
Had a question: if the user makes a mistake on the tag entered during the "install" (say a spelling mistake), can they use the "enroll" command to rewrite the tags for this agent?  (this is assuming that we don't have the ability in the UI yet to modify)

File: internal/pkg/agent/cmd/enroll.go [hunk 3]
```diff
@@ -223,6 +224,10 @@ func buildEnrollmentFlags(cmd *cobra.Command, url string, token string) []string
 		args = append(args, "--fleet-server-es-insecure")
 	}
 
+	for _, v := range fTags {
+		args = append(args, "--tag")
+		args = append(args, v)
```
Comment (michel-laterman) [inline internal/pkg/agent/cmd/enroll.go:229]:
```suggestion
                args = append(args, "--tag", v)
```

File: internal/pkg/agent/cmd/enroll_cmd.go [hunk 4]
```diff
@@ -1027,3 +1030,13 @@ func expBackoffWithContext(ctx context.Context, init, max time.Duration) backoff
 	}()
 	return bo
 }
+
+func cleanTags(tags []string) []string {
+	var r []string
+	for _, str := range tags {
+		if str != "" {
```
Comment (michel-laterman) [inline internal/pkg/agent/cmd/enroll_cmd.go:1037]: Should we run it through `TrimSpace` before this to ensure a tag is not `" "`?

**Review by michel-laterman — APPROVED** [review]
lgtm

**Pull request merged**